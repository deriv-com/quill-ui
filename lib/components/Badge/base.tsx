import { forwardRef } from "react";
import { BadgeProps } from "./types";
import clsx from "clsx";
import "./badge.scss";
import { CaptionText, Text } from "@components/Typography";

export const BadgeBase = forwardRef<HTMLSpanElement, BadgeProps>(
    (
        { size = "sm", color = "success", label, variant, className, ...rest },
        ref,
    ) => {
        const labelColor = "badge__base-color";
        return (
            <span
                className={clsx(
                    className,
                    "badge__base",
                    `badge__variant__${variant}--size-${size}`,
                    `badge__color--${color}`,
                )}
                ref={ref}
                {...rest}
            >
                {label &&
                    (size === "sm" ? (
                        <CaptionText bold color={labelColor}>
                            {label}
                        </CaptionText>
                    ) : size === "md" ? (
                        <Text bold color={labelColor}>
                            {label}
                        </Text>
                    ) : size === "lg" ? (
                        <Text size="xl" bold color={labelColor}>
                            {label}
                        </Text>
                    ) : null)}
            </span>
        );
    },
);
export default BadgeBase;
