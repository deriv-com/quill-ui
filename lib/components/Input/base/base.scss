.quill-input {
    &__container {
        display: flex;
        width: 100%;
        flex-direction: column;
        gap: var(--component-field-spacing-gap-xs);

        .message__container {
            display: flex;
            gap: var(--component-field-spacing-gap-sm);
            justify-content: space-between;
            &--md {
                padding-inline: var(--component-field-spacing-padding-md);
            }
            &--sm {
                padding-inline: var(--component-field-spacing-padding-sm);
            }

            &__status {
                &--neutral {
                    color: var(--component-textIcon-normal-subtle);
                }
                &--success {
                    color: var(--component-textIcon-statusNormal-success);
                }
                &--error {
                    color: var(--component-textIcon-statusNormal-danger);
                }
            }
            &__disabled {
                color: var(--component-textIcon-normal-subtle);
            }

            &__text {
                font-size: var(
                    --semantic-typography-caption-regular-default-fontSize
                );
                line-height: var(
                    --semantic-typography-caption-regular-default-lineHeight
                );
                font-weight: var(
                    --semantic-typography-caption-regular-default-fontWeight
                );
                font-family: var(
                    --semantic-typography-caption-regular-default-fontFamily
                );
            }
        }
        .icon_wrapper {
            display: flex;
            margin: auto;
        }
    }

    &-icons__wrapper {
        display: flex;
        width: 100%;
        gap: var(--component-field-spacing-gap-sm);
    }
    &__rotate {
        transition-property: transform;
        transition-timing-function: var(--core-motion-ease-400);
        transition-duration: var(--core-motion-duration-200);
        transform: rotate(0);

        &[data-state="open"] {
            transform: rotate(180deg);
        }
    }
    &__icon {
        margin: auto;
        fill: var(--component-textIcon-normal-prominent);
    }

    &-label__wrapper {
        display: flex;
        position: relative;
        flex-grow: 1;

        &:has(input[type="select"]) {
            z-index: -1;
        }

        &:focus-within {
            &:has(label) {
                padding-top: var(--core-spacing-600);
            }

            & > .label {
                top: var(--core-spacing-50);
                &__status--neutral {
                    color: var(--component-field-label-color-focus);
                }
            }
        }

        &--has-value {
            &:has(label) {
                top: var(--core-spacing-50);
                padding-top: var(--core-spacing-600);
            }

            & > .label {
                top: var(--core-spacing-50);
            }
        }
    }

    &-no-label__wrapper {
        display: flex;
        width: 100%;
    }

    &-label__label {
        margin: auto var(--core-spacing-200) auto;

        &--right {
            margin-left: var(--core-spacing-200);
        }
    }

    &__wrapper {
        width: 100%;
        align-items: center;
        font-family: var(--core-fontFamily-ibmPlex-sans);
        overflow: hidden;
        position: relative;
        display: flex;
        box-sizing: border-box;
        -moz-box-sizing: border-box;
        -webkit-box-sizing: border-box;

        &:has(input:disabled) {
            cursor: not-allowed;

            .label {
                color: var(--component-field-label-color-disabled);
            }

            svg {
                fill: var(--component-field-label-color-disabled);
            }
        }
        &:hover {
            transition-property: border;
            transition-timing-function: var(--core-motion-ease-400);
            transition-duration: var(--core-motion-duration-200);
        }
        &__variant {
            &--outline {
                z-index: 1;
                background-color: var(--component-field-bg-outline-default);
                border: var(--component-field-border-width) solid
                    var(--component-field-border-color-default);

                &:has(input:disabled) {
                    border-color: var(--component-field-border-color-default);
                }

                &--neutral {
                    &:hover {
                        border-color: var(--component-field-border-color-hover);
                    }
                    &:focus-within,
                    &:focus {
                        border-color: var(--component-field-border-color-focus);
                    }
                }

                &--success {
                    border-color: var(
                        --component-field-border-color-successDefault
                    );
                    &:hover {
                        border-color: var(
                            --component-field-border-color-successHover
                        );
                    }
                    &:focus-within,
                    &:focus {
                        border-color: var(
                            --component-field-border-color-successFocus
                        );
                    }
                }

                &--error {
                    border-color: var(
                        --component-field-border-color-failDefault
                    );
                    &:hover {
                        border-color: var(
                            --component-field-border-color-failHover
                        );
                    }
                    &:focus-within,
                    &:focus {
                        border-color: var(
                            --component-field-border-color-failFocus
                        );
                    }
                }
            }

            &--fill {
                border: var(--component-field-border-width) solid transparent;
                &:has(input:disabled) {
                    background-color: var(--component-field-bg-fill-default);
                }

                &--neutral {
                    background-color: var(--component-field-bg-fill-default);
                    &:hover {
                        background-color: var(--component-field-bg-fill-hover);
                    }
                    &:focus-within,
                    &:focus {
                        border: var(--component-field-border-width) solid
                            var(--component-field-border-color-focus);
                        background-color: var(--component-field-bg-fill-focus);
                    }
                }

                &--success {
                    background-color: var(
                        --component-field-bg-fill-successDefault
                    );
                    &:hover {
                        background-color: var(
                            --component-field-bg-fill-successHover
                        );
                    }
                    &:focus-within,
                    &:focus {
                        border: var(--component-field-border-width) solid
                            var(--component-field-border-color-successFocus);
                        background-color: var(
                            --component-field-bg-fill-successFocus
                        );
                    }
                }

                &--error {
                    background-color: var(
                        --component-field-bg-fill-failDefault
                    );
                    &:hover {
                        background-color: var(
                            --component-field-bg-fill-failHover
                        );
                    }
                    &:focus-within,
                    &:focus {
                        border: var(--component-field-border-width) solid
                            var(--component-field-border-color-failFocus);
                        background-color: var(
                            --component-field-bg-fill-failFocus
                        );
                    }
                }
            }
        }

        &__size {
            &--sm {
                padding-inline: var(--component-field-spacing-padding-sm);
                height: var(--component-field-height-sm);
                border-radius: var(--component-field-border-radius-sm);
                gap: var(--component-field-spacing-gap-sm);
            }
            &--md {
                padding-inline: var(--component-field-spacing-padding-md);
                height: var(--component-field-height-md);
                border-radius: var(--component-field-border-radius-md);
                gap: var(--core-spacing-800);
            }
        }

        &-with_button {
            &-right {
                &--sm {
                    padding-inline-end: var(--core-spacing-200);
                    padding-block: var(--core-spacing-200);
                }
                &--md {
                    padding-inline-end: var(--core-spacing-600);
                    padding-block: var(--core-spacing-600);
                }
            }

            &-bottom {
                &--md {
                    flex-direction: column;
                    height: auto;
                    padding: var(--component-field-spacing-padding-md);

                    &--has-label {
                        gap: var(--core-spacing-1200);
                    }

                    &--has-label:has(input:focus),
                    &--has-value {
                        padding-top: var(--temp-static-spacing-7);
                        gap: var(--temp-static-spacing-15);

                        .input {
                            padding-top: var(--core-spacing-300);
                        }
                        .label {
                            top: var(--temp-static-spacing-7);
                        }
                    }

                    &--no-label {
                        padding: var(--component-field-spacing-padding-md);
                        gap: var(--component-field-spacing-padding-md);

                        .input {
                            padding-top: var(--core-spacing-50);
                        }
                    }
                }
                &--sm {
                    flex-direction: column;
                    height: auto;
                    padding-block: var(--core-spacing-200)
                        var(--core-spacing-400);
                    padding-inline: var(--core-spacing-400);
                }
            }
        }

        .input {
            width: 100%;
            background-color: transparent;
            border: none;
            cursor: pointer;
            &:disabled {
                cursor: not-allowed;
            }
            &:disabled + .label {
                cursor: not-allowed;
            }
            &:focus {
                outline: 2px solid transparent;
                outline-offset: 2px;
                color: var(--component-field-label-color-focus);
            }
            &:hover {
                color: var(--component-field-label-color-default);
            }
            &:placeholder-shown {
                color: var(--component-field-placeholder-color);
            }
            &:not(placeholder-shown) {
                color: var(--component-field-label-color-default);
            }
            &:disabled {
                color: var(--component-field-label-color-disabled);
            }

            &:has(~ label) {
                font-size: var(--core-fontSize-100);
                &::placeholder {
                    color: transparent;
                }
                &:focus {
                    &::placeholder {
                        color: var(--component-field-label-color-disabled);
                    }
                }
            }

            &::-webkit-search-decoration,
            &::-webkit-search-cancel-button,
            &::-webkit-search-results-button,
            &::-webkit-search-results-decoration,
            &::-ms-clear,
            &::-ms-reveal {
                display: none;
            }
            &::-ms-clear,
            &::-ms-reveal {
                width: 0;
                height: 0;
            }
            &__align {
                &--left {
                    text-align: left;
                }
                &--center {
                    &:not(:has(~ label)) {
                        text-align: center;
                    }
                }
            }
            &__size {
                &--sm {
                    font-size: var(
                        --semantic-typography-body-sm-regular-default-fontSize
                    );
                    line-height: var(
                        --semantic-typography-body-sm-regular-default-lineHeight
                    );
                }
                &--md {
                    font-size: var(
                        --semantic-typography-body-md-regular-default-fontSize
                    );
                    line-height: var(
                        --semantic-typography-body-md-regular-default-lineHeight
                    );
                }
            }
        }
        .label {
            position: absolute;
            transition-property: all;
            transition-timing-function: var(--core-motion-ease-400);
            transition-duration: var(--core-motion-duration-200);
            display: flex;
            align-items: center;
            width: 100%;
            cursor: pointer;
            left: var(--core-spacing-50);
            gap: var(--semantic-spacing-gap-sm);

            &__status {
                &--neutral {
                    color: var(--component-field-label-color-default);
                }
                &--success {
                    color: var(--component-field-label-color-success);
                }
                &--error {
                    color: var(--component-field-label-color-fail);
                }
            }

            &-field-marker {
                &__required {
                    &--true {
                        color: var(--core-color-solid-red-900);
                    }

                    &--false {
                        color: var(--component-textIcon-normal-disabled);
                    }
                }
            }
        }
        .peer {
            &:placeholder-shown + .label {
                line-height: var(--core-lineHeight-300);
                font-size: var(--core-fontSize-100);
                height: 100%;
            }
            &:focus + .label {
                line-height: var(--core-lineHeight-100);
                font-size: var(--core-fontSize-50);
                height: var(--core-size-600);
                gap: var(--semantic-spacing-gap-sm);

                &-field-marker {
                    font-size: var(
                        --semantic-typography-caption-regular-default-fontSize
                    );
                }
            }
            &:not(placeholder-shown) + .label {
                line-height: var(--core-lineHeight-100);
                font-size: var(--core-fontSize-50);
                height: var(--core-size-600);

                &-field-marker {
                    font-size: var(
                        --semantic-typography-body-md-regular-default-fontSize
                    );
                }
            }
        }

        .status-icon {
            &--neutral {
                fill: var(--component-textIcon-normal-default);
            }
            &--success {
                fill: var(--component-textIcon-statusNormal-success);
            }
            &--error {
                fill: var(--component-textIcon-statusNormal-danger);
            }
        }
    }
}
