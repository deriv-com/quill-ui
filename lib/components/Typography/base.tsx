import React, { ComponentProps, ElementType } from "react";
import { TGenericSizes } from "@types";
import "@quill/quill.scss";
import "@styles/index.scss";
import "./typography.scss";

export interface TypographyProps extends ComponentProps<THeaders> {
    size?: Extract<TGenericSizes, "lg" | "md" | "sm" | "xl">;
    as?: ElementType | keyof THeaders;
    bold?: boolean;
    italic?: boolean;
    underlined?: boolean;
    color?: string;
    centered?: boolean;
}

type THeaders = "h1" | "h2" | "h3" | "h4" | "h5" | "h6";

export const Typography = ({
    children,
    as = "p",
    ...rest
}: TypographyProps) => {
    const tag = typeof as === "string" ? as : undefined;

    return React.createElement(
        tag || "p",
        {
            ...rest,
        },
        children,
    );
};
