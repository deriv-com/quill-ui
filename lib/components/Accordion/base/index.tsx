import { useCallback, useEffect, useRef, useState } from "react";
import clsx from "clsx";
import { Text } from "@components/Typography";
import { StandaloneChevronDownRegularIcon } from "@deriv/quill-icons/Standalone";
import { AccordionProps } from "../types";
import "./accordion-base.scss";

export const Base = ({
    id,
    className,
    title = "",
    subtitle,
    content: Content,
    expanded = false,
    textSize = "md",
    icon,
    divider = "none",
    disabled = false,
    customContent: CustomContent,
    contentClassname,
    onExpand,
}: AccordionProps) => {
    const [isExpanded, setExpanded] = useState(expanded);
    const [isAutoExpand, setAutoExpand] = useState(false);

    const accordionElement = useRef<HTMLDivElement>(null);

    const toggleCollapse = useCallback(() => {
        setExpanded((current) => !current);
        setAutoExpand(false);
        scrollToExpanded(500);

        if (onExpand) {
            onExpand(!isExpanded, title);
        }
    }, [isExpanded, onExpand, title]);

    const handleKeyUp = useCallback(
        (e: KeyboardEvent) => {
            if (e.code === "Enter" || e.key === "Enter") {
                if (accordionElement.current === document.activeElement) {
                    toggleCollapse();
                }
            }
        },
        [toggleCollapse],
    );

    const scrollToExpanded = (delay = 1000) => {
        const accElement = accordionElement.current;

        if (accElement) {
            setTimeout(() => {
                accElement.scrollIntoView({
                    block: "center",
                    behavior: "smooth",
                });
            }, delay);
        }
    };

    useEffect(() => {
        const hashWithoutSymbol =
            typeof window !== "undefined" && window.location.hash.slice(1);

        if (hashWithoutSymbol === id) {
            setAutoExpand(true);
            scrollToExpanded();
        }
    }, [id]);

    // Key handlers
    useEffect(() => {
        document.addEventListener("keyup", handleKeyUp);
        return () => {
            document.removeEventListener("keyup", handleKeyUp);
        };
    }, [handleKeyUp]);

    useEffect(() => {
        setExpanded(expanded);
    }, [expanded]);

    return (
        <div
            data-id={id}
            ref={accordionElement}
            tabIndex={0}
            className={clsx(
                "quill-accordion-container",
                `quill-accordion-container-divider-${divider}`,
                disabled && `quill-accordion-container-${disabled}`,
                className,
                isExpanded && "quill-accordion-container-expanded-color",
            )}
        >
            <div
                className={clsx(
                    "quill-accordion-base",
                    disabled && "quill-accordion-base-disabled",
                    contentClassname,
                )}
                onClick={() => toggleCollapse()}
                data-testid="toggle-expand"
            >
                {CustomContent ? (
                    <CustomContent />
                ) : (
                    <>
                        {icon && (
                            <div
                                className={clsx(
                                    disabled && "quill-accordion-icon-disabled",
                                )}
                                data-testid="accordion-icon"
                            >
                                {icon}
                            </div>
                        )}
                        <div className={"quill-accordion-header"}>
                            <Text
                                className={clsx(
                                    "quill-accordion-base-header-title-color",
                                    `quill-accordion-base-header-title-${textSize}`,
                                    disabled &&
                                        "quill-accordion-header-title-disabled",
                                )}
                            >
                                {title}
                            </Text>
                            {subtitle && (
                                <Text
                                    className={clsx(
                                        "quill-accordion-header-subtitle-color",
                                        `quill-accordion-base-header-subtitle-${textSize}`,
                                        disabled &&
                                            "quill-accordion-header-subtitle-disabled",
                                    )}
                                >
                                    {subtitle}
                                </Text>
                            )}
                        </div>
                    </>
                )}
                <div
                    className="quill-accordion-icon-rotate"
                    data-state={isExpanded || isAutoExpand ? "open" : "close"}
                    data-testid="chevron"
                >
                    <StandaloneChevronDownRegularIcon
                        iconSize="sm"
                        fill={
                            disabled
                                ? "var(--component-accordion-icon-disabled)"
                                : "var(--component-textIcon-normal-prominent)"
                        }
                    />
                </div>
            </div>

            <div
                className={clsx(
                    disabled && "quill-accordion-disabled",
                    isAutoExpand || isExpanded
                        ? "quill-accordion-expanded"
                        : "quill-accordion-expanded-visible",
                )}
                data-testid="expanded-content"
            >
                <div className={clsx("quill-accordion-expanded-content")}>
                    {Content && <Content />}
                </div>
            </div>
        </div>
    );
};

Base.displayName = "AccordionBase";

export default Base;
