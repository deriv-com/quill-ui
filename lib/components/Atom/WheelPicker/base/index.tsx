import { useEffect, useMemo, useRef, useState } from "react";
import clsx from "clsx";
import { DropdownItem } from "@components/Atom/dropdown";
import "./wheel-picker-base.scss";
import { WheelPickerProps } from "../types";

export const WheelPicker = ({
    data,
    selectedValue,
    setSelectedValue,
    handleKeyDown,
    isFocused,
    dropDownItemClassName,
    containerClassName,
    listClassName,
    position,
    containerHeight = "100%",
    ...rest
}: WheelPickerProps) => {
    const itemsRefs = useRef<HTMLDivElement[]>([]);
    const [inputData, setInputData] = useState(data);
    const dataItemsMap = useMemo(
        () =>
            data.reduce(
                (map, item, index) => map.set(item.value, index),
                new Map(),
            ),
        [data],
    );

    const [currentDataValue, setCurrentValue] = useState(
        dataItemsMap.get(selectedValue) ?? 0,
    );
    const itemsContainerRef = useRef<HTMLUListElement>(null);

    const handleDataScroll = (event: Event) => {
        if (!event.target) return;
        const scrollTop = Math.max(
            (event.target as HTMLUListElement).scrollTop,
            0,
        );
        const selectedElement = Math.min(
            Math.max(Math.floor(scrollTop / 48), 0),
            inputData.length - 1,
        );

        (itemsRefs.current[selectedElement] as HTMLDivElement)?.focus({
            preventScroll: true,
        });
        setSelectedValue(inputData[selectedElement].value);
    };

    const resizeObserver = new ResizeObserver(() => {
        const index = Number(currentDataValue);
        if (!itemsContainerRef?.current) return;
        const divHeight = itemsContainerRef?.current.clientHeight;
        itemsContainerRef.current.style.paddingTop = `${divHeight * 0.5}px`;
        itemsContainerRef.current.style.paddingBottom = `${divHeight * 0.5}px`;
        if (itemsRefs.current[index]) {
            (itemsRefs.current[index] as HTMLDivElement).scrollIntoView({
                block: "center",
            });
        }
    });

    useEffect(() => {
        if (!itemsContainerRef.current) return;
        setSelectedValue(inputData[currentDataValue]?.value);
        resizeObserver.observe(itemsContainerRef.current);
        itemsContainerRef.current.addEventListener("scroll", handleDataScroll);
        return () => {
            resizeObserver.disconnect();
            itemsContainerRef.current?.removeEventListener(
                "scroll",
                handleDataScroll,
            );
        };
    }, [inputData]);

    useEffect(() => {
        setInputData(data);
        setCurrentValue(dataItemsMap.get(selectedValue) ?? 0);
    }, [data]);

    useEffect(() => {
        if (isFocused) {
            itemsRefs.current[dataItemsMap.get(selectedValue)]?.focus();
        }
    }, [isFocused]);

    return (
        <div
            className={clsx(
                "quill-wheel-picker__container",
                `quill-wheel-picker__container-disabled-${rest?.disabled}`,
                containerClassName,
            )}
            style={{ height: containerHeight }}
        >
            <ul
                className={clsx(
                    "quill-wheel-picker__data-items",
                    `quill-wheel-picker__data-items-${position}`,
                    listClassName,
                )}
                role="listbox"
                ref={itemsContainerRef}
                {...rest}
            >
                {inputData.map(({ value, label }, index) => {
                    return (
                        <DropdownItem
                            className={clsx(
                                "quill-wheel-picker__data-item",
                                dropDownItemClassName,
                            )}
                            tabIndex={0}
                            label={label ?? value}
                            role="option"
                            textAlignment="center"
                            ref={(node: HTMLDivElement) =>
                                (itemsRefs.current[index] = node)
                            }
                            key={value}
                            onKeyDown={handleKeyDown}
                            onClick={() => {
                                itemsRefs.current[index]?.scrollIntoView({
                                    block: "center",
                                    behavior: "smooth",
                                    inline: "nearest",
                                });
                                itemsRefs.current[index]?.focus();
                            }}
                            disabled={inputData[index].value !== selectedValue}
                        >
                            {value}
                        </DropdownItem>
                    );
                })}
            </ul>
        </div>
    );
};

export default WheelPicker;
