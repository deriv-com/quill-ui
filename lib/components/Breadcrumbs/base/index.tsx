import { BreadcrumbProps, LinkProps } from "../types";
import { Fragment, useEffect, useState } from "react";
import { maxLinks } from "./constants";
import "./breadcrumbs.scss";
import useBreakpoints from "@hooks/useBreakpoints";
import clsx from "clsx";
import Link from "@components/Link";

export function Base({ size = "sm", links = [], className }: BreadcrumbProps) {
    const [renderLinks, setRenderLinks] = useState<LinkProps[]>([]);
    const [dropdownLinks, setDropdownLinks] = useState<LinkProps[]>([]);

    const isLastItem = (key: number) => renderLinks.length - 1 > key;
    const linksLen = links.length;
    const { isMobile } = useBreakpoints();
    const hasExtra = isMobile && maxLinks < linksLen;

    useEffect(() => {
        setRenderLinks(links);

        if (hasExtra) {
            const limitedLinks = [
                links[0],
                { content: "...", href: "" },
                links[linksLen - 2],
                links[linksLen - 1],
            ];

            setRenderLinks(limitedLinks);

            const remainingLinks: LinkProps[] = links.filter(
                (item) => !new Set(limitedLinks).has(item),
            );

            setDropdownLinks(remainingLinks);
        }
    }, [links]);

    return (
        <div className={clsx("breadcrumb--container", className)}>
            {renderLinks.map(({ content, href }, lk) => (
                <Fragment key={`breadcrumbs-${content}`}>
                    {hasExtra && lk === 1 && (
                        <select
                            className="breadcrumb--dropdown"
                            onChange={(
                                e: React.ChangeEvent<HTMLSelectElement>,
                            ) => {
                                if (typeof window !== "undefined") {
                                    const url = e.target.value;

                                    try {
                                        const sanitizedUrl = new URL(url);
                                        window.location.href =
                                            sanitizedUrl.href;
                                    } catch (err) {
                                        console.error("Invalid URL:", url);
                                    }
                                }
                            }}
                        >
                            {dropdownLinks.map(({ href, content }) => {
                                return (
                                    <option
                                        key={`dropdown-${content}`}
                                        value={href}
                                    >
                                        {content}
                                    </option>
                                );
                            })}
                        </select>
                    )}

                    <Link
                        size={size}
                        href={href}
                        hasHoverEffect={false}
                        hasChevron={isLastItem(lk)}
                        color="black"
                        disabled={!isLastItem(lk)}
                        className="breadcrumb--item"
                    >
                        {content}
                    </Link>
                </Fragment>
            ))}
        </div>
    );
}

export default Base;
