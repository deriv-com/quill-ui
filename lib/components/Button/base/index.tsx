import { forwardRef } from "react";
import clsx from "clsx";
import { ButtonProps } from "../types";
import {
    LabelPairedChevronDownLgBoldIcon,
    LabelPairedChevronDownMdBoldIcon,
    LabelPairedChevronDownSmBoldIcon,
    LabelPairedChevronDownXlBoldIcon,
    LabelPairedLoaderCaptionRegularIcon,
    LabelPairedLoaderLgRegularIcon,
    LabelPairedLoaderMdRegularIcon,
    LabelPairedLoaderSmRegularIcon,
} from "@deriv/quill-icons";
import "../button.scss";
import { CaptionText, Text } from "@components/Typography";

export const ButtonSize = {
    xl: "quill-button__size--xl",
    lg: "quill-button__size--lg",
    md: "quill-button__size--md",
    sm: "quill-button__size--sm",
} as const;

const dropdownIcons = {
    sm: LabelPairedChevronDownSmBoldIcon,
    md: LabelPairedChevronDownMdBoldIcon,
    lg: LabelPairedChevronDownLgBoldIcon,
    xl: LabelPairedChevronDownXlBoldIcon,
};
export const loaderIcons = {
    sm: LabelPairedLoaderCaptionRegularIcon,
    md: LabelPairedLoaderSmRegularIcon,
    lg: LabelPairedLoaderMdRegularIcon,
    xl: LabelPairedLoaderLgRegularIcon,
};

export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
    (
        {
            className,
            color = "coral",
            icon,
            children,
            isDropdownOpen,
            dropdown = false,
            selected,
            fullWidth = false,
            isLoading = false,
            size = "md",
            label,
            disabled,
            iconPosition,
            variant = "primary",
            ...rest
        },
        ref,
    ) => {
        const buttonColorClass = `quill__color--${variant}-${color}`;
        const labelSize = size === "md" ? "sm" : size === "lg" ? "md" : "xl";
        const DropdownIcon = dropdownIcons[size];
        const LoaderIcon = loaderIcons[size];

        return (
            <button
                className={clsx(
                    "quill-button",
                    ButtonSize[size],
                    buttonColorClass,
                    className,
                    fullWidth && "quill-button__full-width",
                )}
                disabled={disabled}
                data-state={selected ? "selected" : ""}
                ref={ref}
                {...rest}
            >
                {iconPosition === "start" && icon && !isLoading && icon}

                {isLoading && !disabled && (
                    <LoaderIcon
                        data-testid="button-loader"
                        className="quill-button__loader-icon"
                    />
                )}
                {children && <div>{children}</div>}
                {label && !isLoading && (
                    <span className="quill-button-label">
                        {size === "sm" ? (
                            <CaptionText color={color} bold>
                                {label}
                            </CaptionText>
                        ) : (
                            <Text size={labelSize} bold color={color}>
                                {label}
                            </Text>
                        )}
                    </span>
                )}
                {iconPosition === "end" && icon && !isLoading && icon}
                {dropdown && DropdownIcon && !isLoading && (
                    <DropdownIcon
                        data-state={isDropdownOpen ? "open" : "close"}
                        className={clsx(
                            "quill-button__transform",
                            isDropdownOpen && "quill-button__transform-rotate",
                        )}
                    />
                )}
            </button>
        );
    },
);

Button.displayName = "Button";

export default Button;
