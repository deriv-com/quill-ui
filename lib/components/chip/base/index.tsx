import React, { useState } from "react";
import { SelectableChipProps } from "../types";
import { StandardSizes } from "../../../types";
import {
    StandaloneChevronDownRegularIcon,
    StandaloneCircleXmarkRegularIcon,
} from "@deriv/quill-icons";
import "../_chip.scss";
import clsx from "clsx";

export const ChipIconSizes: Record<
    StandardSizes,
    { width: number; height: number }
> = {
    sm: {
        width: 11,
        height: 18,
    },
    md: {
        width: 13,
        height: 22,
    },
    lg: {
        width: 14,
        height: 24,
    },
};

export const Base = ({
    icon: Icon,
    size = "md",
    children,
    labelTag,
    dismissible = false,
    dropdown = false,
    className,
    onChipSelect,
    onDismiss,
    ...rest
}: SelectableChipProps) => {
    const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
        // if (dismissible || dropdown) return;
        const target = event.currentTarget;
        const icon = target.querySelector(".rotate");
        const isOpen = icon?.getAttribute("data-state") === "open";
        const selected_state = isOpen ? "close" : "open";
        icon?.setAttribute("data-state", selected_state);
        onChipSelect?.(event, !isOpen);
    };

    const handleDismiss = (
        event: React.MouseEvent<SVGSVGElement, MouseEvent>,
    ) => {
        onDismiss?.(event);
    };

    return (
        <button
            onClick={handleClick}
            className={clsx(`quill-chip__size--${size}`, className)}
            {...rest}
        >
            {Icon && <Icon {...ChipIconSizes[size]} />}
            {children}
            {labelTag && <span className="">{labelTag}</span>}
            {dismissible && (
                <StandaloneCircleXmarkRegularIcon
                    {...ChipIconSizes[size]}
                    onClick={handleDismiss}
                    data-testid="dt-chip-dismissible-btn"
                    className="cursor-pointer"
                />
            )}
            {dropdown && (
                <StandaloneChevronDownRegularIcon
                    width={24}
                    height={24}
                    data-state="open"
                    className="rotate"
                />
            )}
        </button>
    );
};

export default Base;
