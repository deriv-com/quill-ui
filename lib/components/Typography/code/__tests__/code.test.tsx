import { render, screen } from "@testing-library/react";
import { CodeText } from "..";
import { TypographyProps } from "../../base";

const captionSizes: TypographyProps["size"][] = ["sm", "md", "lg", "xl"];

describe("CaptionText", () => {
    it("should display correct content", () => {
        render(<CodeText>CodeText content</CodeText>);

        const captionText = screen.getByText(/CodeText content/);
        expect(captionText).toBeInTheDocument();
    });

    it("should have correct classnames with default values", () => {
        const { container } = render(<CodeText>CodeText content</CodeText>);

        expect(container).toMatchSnapshot();
    });

    it("should have correct classnames with bold enabled", () => {
        const { container } = render(
            <CodeText bold>CaptionText content</CodeText>,
        );

        expect(container).toMatchSnapshot();
    });

    it("should have correct classnames with italic enabled", () => {
        const { container } = render(
            <CodeText italic>CaptionText content</CodeText>,
        );

        expect(container).toMatchSnapshot();
    });

    it("should have correct classnames with underline enabled", () => {
        const { container } = render(
            <CodeText underlined>CaptionText content</CodeText>,
        );

        expect(container).toMatchSnapshot();
    });

    captionSizes.forEach((size) => {
        it(`should have correct classnames with size ${size}`, () => {
            const { container } = render(
                <CodeText size={size}>CodeText content</CodeText>,
            );

            expect(container).toMatchSnapshot();
        });
    });
});
