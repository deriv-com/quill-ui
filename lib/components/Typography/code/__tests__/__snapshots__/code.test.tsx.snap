// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CaptionText should have correct classnames with bold enabled 1`] = `
<div>
  <code
    class="quill-typography__code-text__size--md__weight--bold__decoration--default quill-typography__color--default"
  >
    CaptionText content
  </code>
</div>
`;

exports[`CaptionText should have correct classnames with default values 1`] = `
<div>
  <code
    class="quill-typography__code-text__size--md__weight--regular__decoration--default quill-typography__color--default"
  >
    CodeText content
  </code>
</div>
`;

exports[`CaptionText should have correct classnames with italic enabled 1`] = `
<div>
  <code
    class="quill-typography__code-text__size--md__weight--regular__decoration--italic quill-typography__color--default"
  >
    CaptionText content
  </code>
</div>
`;

exports[`CaptionText should have correct classnames with size lg 1`] = `
<div>
  <code
    class="quill-typography__code-text__size--lg__weight--regular__decoration--default quill-typography__color--default"
  >
    CodeText content
  </code>
</div>
`;

exports[`CaptionText should have correct classnames with size md 1`] = `
<div>
  <code
    class="quill-typography__code-text__size--md__weight--regular__decoration--default quill-typography__color--default"
  >
    CodeText content
  </code>
</div>
`;

exports[`CaptionText should have correct classnames with size sm 1`] = `
<div>
  <code
    class="quill-typography__code-text__size--sm__weight--regular__decoration--default quill-typography__color--default"
  >
    CodeText content
  </code>
</div>
`;

exports[`CaptionText should have correct classnames with size xl 1`] = `
<div>
  <code
    class="quill-typography__code-text__size--xl__weight--regular__decoration--default quill-typography__color--default"
  >
    CodeText content
  </code>
</div>
`;

exports[`CaptionText should have correct classnames with underline enabled 1`] = `
<div>
  <code
    class="quill-typography__code-text__size--md__weight--regular__decoration--underline quill-typography__color--default"
  >
    CaptionText content
  </code>
</div>
`;
